compiler.rb

```diff
@@ -86,32 +86,36 @@
 end.parse!
 # rubocop:enable Metrics/BlockLength
 
-raise 'Cannot use -p/--products and -a/--all simultaneously' if product_names && all_products
-raise 'Either -p/--products OR -a/--all must be present' if product_names.nil? && !all_products
+raise 'Cannot use -p/--products and -a/--all simultaneously' \
+  if products_to_compile && all_products
+raise 'Either -p/--products OR -a/--all must be present' \
+  if products_to_compile.nil? && !all_products
 raise 'Option -o/--output is a required parameter' if output_path.nil?
 raise 'Option -e/--engine is a required parameter' if provider_name.nil?
 
-if all_products
-  product_names = []
-  Dir['products/**/api.yaml'].each do |file_path|
-    product_names.push(File.dirname(file_path))
-  end
+all_product_files = []
+Dir['products/**/api.yaml'].each do |file_path|
+  all_product_files.push(File.dirname(file_path))
+end
 
-  if override_dir
-    Dir["#{override_dir}/products/**/api.yaml"].each do |file_path|
-      product = File.dirname(Pathname.new(file_path).relative_path_from(override_dir))
-      product_names.push(product) unless product_names.include? product
-    end
+if override_dir
+  Dir["#{override_dir}/products/**/api.yaml"].each do |file_path|
+    product = File.dirname(Pathname.new(file_path).relative_path_from(override_dir))
+    all_product_files.push(product) unless products_to_compile.include? product
   end
+end
 
-  raise 'No api.yaml files found. Check provider/engine name.' if product_names.empty?
+if all_products
+  products_to_compile = all_product_files
+  raise 'No api.yaml files found. Check provider/engine name.' if products_to_compile.empty?
```

Comment:
This exception should live after the conditional, right? If I specified an invalid product and not `all`, I wouldn't fail on it anymore.